### DISCUSSION TITLE
[question] computation of val_wer when validation_ds contains multiple manifest files

### DISCUSSION THREAD

**Discussion by itzsimpl** [NVIDIA-NeMo/NeMo#4025] [Q&A] [answered]
It seems that if one uses multiple manifest files for validation_ds the `val_wer` used in `checkpoint_callback` is based on the first manifest file, or computed as the average over all manifest files. This has an obvious influence on which checkpoints are retained `save_top_k`. Is this intended by design? 

If not; having two manifests, one small and one big will give an unfairly large influence to the small manifest in case of averaging, or exclude one if based on manifest with index 0. I believe the correct approach would be to do a division between the sum of numerators and sum of denominators over individual manifests (hence the same result if the two manifests were merged into one) as mentioned in https://github.com/NVIDIA/NeMo/blob/2aa68822bbfb8e7e8c221817b66de28928d6443b/nemo/collections/asr/metrics/wer.py#L75, but I do not see this happening. 

Browsing through code I can see https://github.com/NVIDIA/NeMo/blob/2aa68822bbfb8e7e8c221817b66de28928d6443b/nemo/collections/asr/models/asr_model.py#L40-L45, but my Tensorboard apart from `val_wer` contains also `<manifestfile>_val_wer` and `val_wer==<manifestfile#0>_val_wer`. This value is also used for checkpoints. I am unable to find where in the code the individual `<manifestfile>_val_wer` are appended.

**Comment by VahidooX** [comment]
The val_wer and saving checkpoints are based one just the first manifest in the list. I don't think we take any average or something. If you want to do something like this, you can sample from different validation manifest and create a new one to be used as the first one.

**Comment by titu1994** [comment, marked as answer]
Note that we particularly avoid average of manifests since it is usually meaningless. There's no baseline to compare against such averaged datasets. 

As Vahid suggested, you can still do it by concatenating all the Dev sets you want into one and then using it as the first but I would recommend not doing that. While wer would appear on average to drop, it does not mean that your average accuracy on unseen test data would also be good on average - some datasets might dominate due to amount of data and the average would still drop, sacrificing the accuracy of the smaller datasets. 

It is simply better to separate all manifests, and focus the model to improve on the one you most care about (and usually it is the more challenging datasets in the mix - somewhat like dev other for LS instead of using dev clean as the first dataset)

**Comment by itzsimpl** [comment]
@titu1994 thank you for the explanation of the reasoning behind it. I have, in the meantime, found  https://github.com/NVIDIA/NeMo/blob/2aa68822bbfb8e7e8c221817b66de28928d6443b/nemo/core/classes/modelPT.py#L637-L660 where the approach is explained. 

My question was spurred by the fact that by reading only the config yaml when I added multiple validation sets (in addition to my principal validation set, I added a very small cherry picked one, but placed it as the first one in the list). I was somewhat expecting an average to be reported and used for checkpointing, surprised by the values I later found out they were based just on the cherry picked set. 

If you don't mind, I would suggest adding `val_dl_idx: 0` with a comment that this is the index of the validation set for reporting and checkpointing to all config yamls. I think this alone would help a lot, already. For example, I wasn't aware such a parameter even existed before reading trough the code.

**Comment by titu1994** [comment]
val_dl_idx is not mentioned for a good reason - it is really problematic the moment you finish training and save a pretrained Nemo file. 

Next time you load it up, and you try to fine-tune the model, say your validation setup no longer has that many manifests (just 1 for example). Now the val_dl_idx remains some non 0 number but doesn't index an appropriate manifest, so after training when you try to save a file, it will crash after a training epoch.

**Comment by titu1994** [comment]
Honestly, this system is outdated and doesnt support the original goal - uniform multi dataloader support across domains. Turns out there's no need for multi data support in NLP, NMT during evaluation since they simply compute loss over text slices, not needing manifests. TTS never needs such validation metrics over multiple datasets. 

So it is actually wrong that such support was added to ModelPT, but for example other domains in the future still may need it so it's stuck around.

**Comment by itzsimpl** [comment]
I understand it is outdated. As I understood, the approach in `test_ds` is the same, am I right? For clarity and ease of use purposes, wouldn't a simple warning work in the case of fine-tuning? I mean in case `validation_ds` is changed on load time, check if `val_dl_ids` differs from 0 and issue a warning in case it does, error if it's out of range? 

I imagined also that NMT could take advantage of a multi-validation set, e.g. sacreBLEU scores for written, spoken, scientific or other sub-domain text, ... am I mistaken?

**Comment by titu1994** [comment]
NMT needed it's own setup for multi validation and it's more domain dependent setup. 

Instead of a warning, you'll just hard crash at the moment and there's no easy way to check at the time of setup of the dataloaders themselves.

**Comment by itzsimpl** [comment]
I see. Thank you for the explanation. I think you can convert the issue to a discussion now, so it can come handy to others.

**Comment by titu1994** [comment]
Good point will do.
